external_parser/multistep_example_joiner.cc

```diff
@@ -0,0 +1,246 @@
+    }
+    default:
+      break;
+  }
+  return true;
+}
+
+void multistep_example_joiner::set_default_reward(float default_reward) {
+  _default_reward = default_reward;
+}
+
+void multistep_example_joiner::set_learning_mode_config(v2::LearningModeType learning_mode) {
+  _learning_mode_config = learning_mode;
+}
+
+void multistep_example_joiner::set_problem_type_config(v2::ProblemType problem_type) {
+  _problem_type_config = problem_type;
+}
+
+void multistep_example_joiner::set_reward_function(const v2::RewardFunctionType type) {
+  switch (type) {
+  case v2::RewardFunctionType_Earliest:
+    _reward_calculation = &reward::earliest;
+    break;
+  case v2::RewardFunctionType_Average:
+    _reward_calculation = &reward::average;
+    break;
+
+  case v2::RewardFunctionType_Sum:
+    _reward_calculation = &reward::sum;
+    break;
+
+  case v2::RewardFunctionType_Min:
+    _reward_calculation = &reward::min;
+    break;
+
+  case v2::RewardFunctionType_Max:
+    _reward_calculation = &reward::max;
+    break;
+
+  case v2::RewardFunctionType_Median:
+    _reward_calculation = &reward::median;
+    break;
+
+  default:
+    break;
+  }
+}
+
+void multistep_example_joiner::populate_order() {
+  //TODO: topological sort
+  for (const auto it: _interactions) {
+    _order.push(it.first);
+  }
+  _sorted = true;
+}
+
+outcome_event multistep_example_joiner::process_outcome(const multistep_example_joiner::Parsed<v2::OutcomeEvent> &event_meta) {
+  const auto& metadata = event_meta.meta;
+  const auto& event = event_meta.event;
+  outcome_event o_event;
+  o_event.metadata = {timestamp_to_chrono(*metadata.client_time_utc()),
+                      metadata.app_id() ? metadata.app_id()->str() : "",
+                      metadata.payload_type(),
+                      metadata.pass_probability(),
+                      metadata.encoding(),
+                      metadata.id()->str()};
+
+  if (event.value_type() == v2::OutcomeValue_literal) {
+    o_event.s_value = event.value_as_literal()->c_str();
+  } else if (event.value_type() == v2::OutcomeValue_numeric) {
+    o_event.value = event.value_as_numeric()->value();
+  }
+
+  return o_event;
+}
+
+joined_event multistep_example_joiner::process_interaction(
+    const multistep_example_joiner::Parsed<v2::MultiStepEvent> &event_meta,
+    v_array<example *> &examples) {
+  const auto& metadata = event_meta.meta;
+  const auto& event = event_meta.event;
+  metadata_info meta = {metadata.client_time_utc()
+                         ? timestamp_to_chrono(*metadata.client_time_utc())
+                         : TimePoint(),
+                        metadata.app_id() ? metadata.app_id()->str() : "",
+                        metadata.payload_type(),
+                        metadata.pass_probability(),
+                        metadata.encoding(),
+                        metadata.id()->str(),
+                        v2::LearningModeType::LearningModeType_Online};
+
+  DecisionServiceInteraction data;
```

Comment:
I wonder if some of this could be moved to the parent class.. maybe as a next step/refactor we can look at what we could refactor out

Reply:
yes, there are number of that kind of improvements that I am planning to do in next PRs